src/Core/ProjectFilesWatcher.cs

```diff
@@ -0,0 +1,203 @@
+                    value.ScriptFiles.CollectionChanged += OnScriptsChanged;
+                }
+            }
+        }
+
+        private void AddScripts(IList newScripts)
+        {
+            foreach(ScriptFile script in newScripts)
+            {
+                var directoryName = Path.GetDirectoryName(script.Filename);
+                var fileName = Path.GetFileName(script.Filename);
+                var watcher = new FileSystemWatcher(directoryName, fileName);
+                if (scriptWatchers.TryAdd(script.Filename, watcher))
+                {
+                    watcher.NotifyFilter = NotifyFilters.LastWrite;
+                    watcher.Changed += OnScriptFileChanged;
+                    watcher.EnableRaisingEvents = true;
+                }
+                else
+                {
+                    watcher.Dispose();
+                }
+            }
+        }
+
+        private void RemoveScripts(IList oldScripts)
+        {
+            foreach (ScriptFile script in oldScripts)
+            {
+                if (scriptWatchers.TryRemove(script.Filename, out var watcher))
+                {
+                    watcher.Changed -= OnScriptFileChanged;
+                    watcher.EnableRaisingEvents = false;
+                    watcher.Dispose();
+                }
+            }
+        }
+
+        private void ReloadScript(ScriptFile scriptFile)
+        {
+            var fileName = scriptFile.Filename;
+            try
+            {
+                var bytes = loader.LoadImageBytes(fileName, 0);
+                var stream = new MemoryStream(bytes);
+                using var rdr = new StreamReader(stream);
+                scriptFile.Evaluate(rdr.ReadToEnd());
+            }
+            catch(Exception ex)
+            {
+                eventListener.Error(ex, $"Failed to load script {fileName}.");
+            }
+        }
+
+        private void OnScriptFileChanged(object sender, FileSystemEventArgs e)
+        {
+            var fullPath = e.FullPath;
+            if (!changedFiles.TryAdd(fullPath, true))
+                return;
+            // Multiple Changed event can be raised in some cases. And file
+            // access error can be occurred when processing first event. Use
+            // timer to delay processing event.
+            var timer = new Timer(200)
+            {
+                AutoReset = false,
+            };
+            timer.Elapsed += (sender, e) =>
+            {
+                timer.Dispose();
+                if (!changedFiles.TryRemove(fullPath, out _))
+                    return;
+                OnScriptFileChanged(Project, fullPath);
+            };
+            timer.Start();
+        }
+
+        private void OnScriptFileChanged(Project? project, string fileName)
+        {
+            if (project is null)
+                return;
+            var fullPath = Path.GetFullPath(fileName);
+            var scriptFile = project.ScriptFiles.ToList().
```

Comment:
ToList() is redundant since `ConcurrentObservableCollection.GetEnumerator()` already makes a  call to `ToList()`